release-23.2: raft: re-enable config change safety

pav-kv opened the pull request:
This PR sets the `release-23.2` branch to use the CRDB fork of `etcd-io/raft`, for porting fixes into it, and cherry-picks cockroachdb/raft/pull/3.

Related to #124804, #124820

Epic: none
Release note: none
Release justification: fixing safety bug

pav-kv committed:
release-23.2: use raft fork

This commit set the release-23.2 branch to use the CRDB fork of
etcd-io/raft, for cherry-picking fixes into it.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125552)
<!-- Reviewable:end -->

pav-kv commented:
@nvanbenschoten Let's merge https://github.com/cockroachdb/raft/pull/3 first, and then I will update this PR to also cherry-pick that fix.

pav-kv committed:
release-23.2: cherry-pick cockroachdb/raft/pull/3

Config changes in this raft implementation require a safety constraint:
the leader must not append a config change if it hasn't applied all
config changes in its log.

The DisableConfChangeValidation flag disables this check under the
assumption that the state machine layer provides the equivalent
guarantee. However, it is hard to argue that this is true in split
leaseholder/leader scenarios.

This commit re-enables this check, to bring the safety back. The other
two state-machine-level checks concerned with entering and leaving joint
configs can still be disabled.

Epic: none
Release note: none

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125552#-:-O-fj4WN7DhHmZcQWUtK:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->
